File: UI/scripts/split_api.py
import os
import re
import sys

def main():
    api_path = r"c:/Users/WELCOME/Desktop/Recruit/UI/lib/api.ts"
    out_dir = r"c:/Users/WELCOME/Desktop/Recruit/UI/lib/api"
    os.makedirs(out_dir, exist_ok=True)
    
    with open(api_path, "r", encoding="utf-8") as f:
        content = f.read()

    # 1. Extract types
    # types are typically: export interface Name { ... } or export type Name = ...;
    # Regex to grab block of type / interface
    # Handling nested brackets is tricky with pure regex, but TypeScript interfaces in api.ts are mostly flat.
    
    type_pattern = re.compile(r'((?:(?:\/\*\*.*?\*\/|//[^\n]*)\s*)*export\s+(?:interface|type)\s+\w+.*?^})', re.MULTILINE | re.DOTALL)
    # The above assumes interface ends with `^}`. If it's a type like `export type ActivityType = "x" | "y";`, it ends with `;`
    
    # Let's use a simpler heuristic. We'll split the file by lines and parse block by block.
    
    blocks = []
    lines = content.split('\n')
    
    current_block = []
    
    # We will just categorize functions by their URLs in fetchApi.
    
    # Let's manually define the new files.
    client_ts = """export const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api/v1";

export async function fetchApi<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  const url = `${API_BASE_URL}${endpoint}`;
  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...options.headers,
    },
    credentials: "omit", // Will be changed to include in step #17
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || "API Request Failed");
  }

  return response.json();
}

export async function fetchApiWithAuth<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  // Try fallback logic
  let token = null;
  if (typeof window !== "undefined") {
    token = localStorage.getItem("auth_token");
  }

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...((options.headers as any) || {}),
  };

  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const url = `${API_BASE_URL}${endpoint}`;
  const response = await fetch(url, {
    ...options,
    headers,
    credentials: "omit", // Will change to include
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || "API Request Failed");
  }

  return response.json();
}
"""

    # For types, it is much easier to just grab all interfaces and types
    type_lines = []
    auth_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]
    courses_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]
    jobs_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]
    payments_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]
    admin_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]
    notifications_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]
    tracking_lines = ["import { fetchApi, fetchApiWithAuth } from './client';", "import * as T from './types';"]

    i = 0
    in_type = False
    in_function = False
    brace_count = 0
    buffer = []
    
    while i < len(lines):
        line = lines[i]
        
        # very basic comment handling
        if line.startswith('// ') or line.startswith('/**'):
            # Grab the whole comment block
            idx = i
            while idx < len(lines) and (lines[idx].strip().startswith('//') or lines[idx].strip().startswith('/*') or lines[idx].strip().startswith('*')):
                buffer.append(lines[idx])
                if lines[idx].strip().endswith('*/'):
                    idx += 1
                    break
                idx += 1
            i = idx
            continue

        if not in_type and not in_function:
            if line.startswith('export interface') or line.startswith('export type'):
                in_type = True
                buffer.append(line)
                brace_count += line.count('{') - line.count('}')
                if brace_count == 0 and ';' in line and not line.endswith('{'):
                    in_type = False
                    type_lines.extend(buffer)
                    buffer = []
            elif line.startswith('export async function') or line.startswith('export function'):
                in_function = True
                buffer.append(line)
                brace_count += line.count('{') - line.count('}')
            elif 'export const mapMaterialToUI = ' in line:
               in_function = True
               buffer.append(line)
               brace_count += line.count('{') - line.count('}')
            else:
                pass # ignore client constants etc
        else:
            buffer.append(line)
            brace_count += line.count('{') - line.count('}')
            if brace_count == 0:
                if in_type:
                    type_lines.extend(buffer)
                    type_lines.append("")
                elif in_function:
                    content_str = "\\n".join(buffer)
                    if 'export const mapMaterialToUI' in content_str:
                         courses_lines.extend(buffer)
                    elif '/auth' in content_str or 'login' in content_str or 'register' in content_str or '/profile' in content_str or '/onboarding' in content_str:
                        auth_lines.extend(buffer)
                    elif '/courses' in content_str or '/students/materials' in content_str:
                        courses_lines.extend(buffer)
                    elif '/jobs' in content_str or '/applications' in content_str:
                        jobs_lines.extend(buffer)
                    elif '/payments' in content_str:
                        payments_lines.extend(buffer)
                    elif '/admin' in content_str:
                        admin_lines.extend(buffer)
                    elif '/notifications' in content_str:
                        notifications_lines.extend(buffer)
                    elif '/tracking' in content_str or 'heartbeat' in content_str:
                        tracking_lines.extend(buffer)
                    else:
                        print(f"Unknown category for function: {buffer[0]}")
                        auth_lines.extend(buffer) # fallback
                    
                    auth_lines.append("")
                    courses_lines.append("")
                    jobs_lines.append("")
                    payments_lines.append("")
                    admin_lines.append("")
                    notifications_lines.append("")
                    tracking_lines.append("")
                
                in_type = False
                in_function = False
                buffer = []
        i += 1

    # In auth_lines, etc... replace "export async function getX(): Promise<Y>" with Promise<T.Y>
    # Also fix T.T if we accidentally over-prefix.
    def prefix_types(lines_list):
        out = []
        for line in lines_list:
            # We want to prefix types used in `Promise<Type>` or `Type[]` or `: Type`
            # This is hard to do perfectly with regex without a full parser,
            # But we only need it to compile.
            # Actually, it's safer to just export everything from types.ts and import * as types? 
            # Or export all types from types.ts, and in each file import what we need.
            # Easiest way in generated files: import * as Types from "./types"; then use standard substitution, but that's risky.
            out.append(line)
        return out

    # Write files
    with open(os.path.join(out_dir, "client.ts"), "w", encoding="utf-8") as f:
        f.write(client_ts)
        
    with open(os.path.join(out_dir, "types.ts"), "w", encoding="utf-8") as f:
        f.write("\\n".join(type_lines))
"""

if __name__ == "__main__":
    print("Script started but paused — manual split might be more accurate using standard TS tools.")

File: UI/scripts/split_api.js
const fs = require('fs');
const path = require('path');

const inFilePath = path.join(__dirname, '../lib/api.ts');
const outDir = path.join(__dirname, '../lib/api');

if (!fs.existsSync(outDir)) {
  fs.mkdirSync(outDir, { recursive: true });
}

const content = fs.readFileSync(inFilePath, 'utf-8');
const lines = content.split('\n');

const clientContent = `export const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000/api/v1";

export async function fetchApi<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  const url = \`\${API_BASE_URL}\${endpoint}\`;
  const response = await fetch(url, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...options.headers,
    },
    credentials: "omit",
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || "API Request Failed");
  }

  return response.json();
}

export async function fetchApiWithAuth<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
  let token = null;
  if (typeof window !== "undefined") {
    token = localStorage.getItem("auth_token");
  }

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    ...((options.headers as any) || {}),
  };

  if (token) {
    headers["Authorization"] = \`Bearer \${token}\`;
  }

  const url = \`\${API_BASE_URL}\${endpoint}\`;
  const response = await fetch(url, {
    ...options,
    headers,
    credentials: "omit",
  });

  if (!response.ok) {
    const errorData = await response.json().catch(() => ({}));
    throw new Error(errorData.detail || "API Request Failed");
  }

  return response.json();
}
`;

const types = [];
const auth = [];
const courses = [];
const jobs = [];
const payments = [];
const admin = [];
const notifications = [];
const tracking = [];

let currentArray = null;
let braceCount = 0;
let buffer = [];
let inBlock = false;
let blockType = null; // 'type', 'func'

const STANDARD_IMPORTS = `import { fetchApi, fetchApiWithAuth } from './client';\nimport * as T from './types';\n\n`;

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  
  // Exclude definitions of constants that are going to client.ts
  if (!inBlock && (line.includes('const API_BASE') || line.includes('function fetchApi'))) {
     continue;
  }

  if (!inBlock) {
    if (line.trim().startsWith('export interface') || line.trim().startsWith('export type')) {
      inBlock = true;
      blockType = 'type';
      buffer.push(line);
      braceCount += (line.match(/\\{/g) || []).length - (line.match(/\\}/g) || []).length;
      if (braceCount === 0 && line.includes(';') && !line.includes('{')) {
        inBlock = false;
        types.push(buffer.join('\\n'));
        buffer = [];
      }
    } else if (line.trim().startsWith('export async function') || line.trim().startsWith('export function') || line.trim().startsWith('export const mapMaterialToUI')) {
      inBlock = true;
      blockType = 'func';
      buffer.push(line);
      braceCount += (line.match(/\\{/g) || []).length - (line.match(/\\}/g) || []).length;
      // if single line function layout
      if (braceCount === 0 && line.includes('}')) {
        inBlock = false;
        categorizeFunc(buffer.join('\\n'));
        buffer = [];
      }
    } else if (line.trim() !== '' && !line.trim().startsWith('/**') && !line.trim().startsWith('*') && !line.trim().startsWith('//') && !line.trim().startsWith('*/') && !line.trim().startsWith('import')) {
       // Could be a generic const we might need, like API_BASE
    }
  } else {
    buffer.push(line);
    braceCount += (line.match(/\\{/g) || []).length - (line.match(/\\}/g) || []).length;
    if (braceCount === 0) {
      inBlock = false;
      const str = buffer.join('\\n');
      if (blockType === 'type') {
         types.push(str);
      } else {
         categorizeFunc(str);
      }
      buffer = [];
    }
  }
}

function categorizeFunc(str) {
  if (str.includes('mapMaterialToUI')) { courses.push(str); return; }
  const s = str.toLowerCase();
  if (s.includes('/auth') || s.includes('login') || s.includes('register') || s.includes('/profile') || s.includes('/onboarding')) {
    auth.push(str);
  } else if (s.includes('/courses') || s.includes('/quizzes') || s.includes('material') || s.includes('studentdashboard') || s.includes('flashcard')) {
    courses.push(str);
  } else if (s.includes('/jobs') || s.includes('/applications') || s.includes('match')) {
    jobs.push(str);
  } else if (s.includes('/payments') || s.includes('order') || s.includes('certificate')) {
    payments.push(str);
  } else if (s.includes('/admin')) {
    admin.push(str);
  } else if (s.includes('/notifications')) {
    notifications.push(str);
  } else if (s.includes('/tracking') || s.includes('session') || s.includes('heartbeat')) {
    tracking.push(str);
  } else {
    auth.push(str); // fallback
  }
}

// Write the files
fs.writeFileSync(path.join(outDir, 'client.ts'), clientContent);
fs.writeFileSync(path.join(outDir, 'types.ts'), types.join('\\n\\n') + '\\n');

function writeModule(name, arr) {
  if (arr.length === 0) return;
  // Replace types like Promise<CourseListItem> with Promise<T.CourseListItem>
  // This is a naive replace but should cover most cases in our api.ts
  let text = arr.join('\\n\\n');
  
  // Use regex to prepend T. to any capitalized word that is used as a Type in signatures.
  // Actually, standard T imports are easier via simple exports: export * as T from './types'
  // But wait, it's inside the same file, we can just replace the interfaces explicitly.
  const allTypeNames = types.map(t => {
     const m = t.match(/export (?:interface|type) ([A-Za-z0-9_]+)/);
     return m ? m[1] : null;
  }).filter(Boolean);

  let newText = text;
  // Add T. prefix
  allTypeNames.forEach(tn => {
     // match word boundaries
     const reg = new RegExp("\\\\b" + tn + "\\\\b", 'g');
     newText = newText.replace(reg, "T." + tn);
  });

  fs.writeFileSync(path.join(outDir, name), STANDARD_IMPORTS + newText + '\\n');
}

writeModule('auth.ts', auth);
writeModule('courses.ts', courses);
writeModule('jobs.ts', jobs);
writeModule('payments.ts', payments);
writeModule('admin.ts', admin);
writeModule('notifications.ts', notifications);
writeModule('tracking.ts', tracking);

// Generate index.ts to re-export everything for backward compatibility
const indexContent = `export * from './client';
export * from './types';
export * from './auth';
export * from './courses';
export * from './jobs';
export * from './payments';
export * from './admin';
export * from './notifications';
export * from './tracking';
`;

fs.writeFileSync(path.join(outDir, 'index.ts'), indexContent);

// Rename original api.ts to api.backup.ts
fs.renameSync(inFilePath, inFilePath.replace('api.ts', 'api.backup.ts'));

// Create a new api.ts that just exports from api/index.ts
fs.writeFileSync(inFilePath, `export * from './api/index';\n`);

console.log('Successfully split api.ts');
